### PR TITLE
Fix for PopOS os_release problem

### PR DESCRIPTION
**Pull request by tyleraharrison**
Fixes issue #276

### CODE REVIEW

**Comment by bobslept** [comment]
Nice! Great improvement for the PopOS users. Maybe before offering this option to the user, check if `/etc/pop-os/os-release` exists so that you are for sure being on the PopOS distro which causes this error?

**Comment by AdnanHodzic** [comment]
> Nice! Great improvement for the PopOS users. Maybe before offering this option to the user, check if `/etc/pop-os/os-release` exists so that you are for sure being on the PopOS distro which causes this error?

Absolutely agree, let's run this only if detected distro is pop os.

**Comment by tyleraharrison** [comment]
How does this look?

File: auto_cpufreq/core.py [hunk 1]
```diff
@@ -78,7 +78,28 @@ def get_config(config_file=''):
...
+    dist_name = distro.id()
+except PermissionError:
+    # Current work-around for distros like Pop!_OS where symlink causes permission issues
+    print("Warning: Cannot get distro name")
+    if os.path.exists("/etc/pop-os/os-release"):
+        print("Pop!_OS detected")
+        print("Pop!_OS uses a symbolic link for the os-release file, this causes issues \
+            and can be fixed by converting to a hard link")
+        print("Attempting to change symlink to hard link for /etc/os-release -> /etc/pop-os/os-release")
+    else:
```
Comment (bobslept) [inline auto_cpufreq/core.py:91]: I would remove these

Comment (bobslept) [inline auto_cpufreq/core.py:91]: I would drop this else block

File: auto_cpufreq/core.py [hunk 2]
```diff
@@ -78,7 +78,28 @@ def get_config(config_file=''):
...
+        print("Aborting...")
+        sys.exit(1)
+    yN = input("Continue? [y/N] ")
```
Comment (bobslept) [inline auto_cpufreq/core.py:94]: And just add this to the os.path.exists if block

**Comment by bobslept** [comment]
Almost there, I would just do it like this: 

```
try:
    dist_name = distro.id()
except PermissionError:
    # Current work-around for distros like Pop!_OS where symlink causes permission issues
    print("Warning: Cannot get distro name")
    if os.path.exists("/etc/pop-os/os-release"):
        print("Pop!_OS detected")
        print("Pop!_OS uses a symbolic link for the os-release file, this causes issues \
            and can be fixed by converting to a hard link")
        print("Attempting to change symlink to hard link for /etc/os-release -> /etc/pop-os/os-release")

        yN = input("Continue? [y/N] ")
        if yN.lower() == "y":
            # Backup /etc/os-release
            os.system("sudo mv /etc/os-release /etc/os-release-backup")
            # Create hard link to /etc/os-release
            os.system("sudo ln /etc/pop-os/os-release /etc/os-release")
    else:
        print("Aborting...")
        sys.exit(1)
```

I've tried to comment the code as review kind of thing on github, but never did that before. So I think this is a little bit easier for now. But I will try to add my comments via review thingy, just as a test.

File: auto_cpufreq/core.py [hunk 3]
```diff
@@ -78,7 +78,28 @@ def get_config(config_file=''):
...
+try:
+    dist_name = distro.id()
+except PermissionError:
+    # Current work-around for distros like Pop!_OS where symlink causes permission issues
+    print("Warning: Cannot get distro name")
+    if os.path.exists("/etc/pop-os/os-release"):
+        print("Pop!_OS detected")
+        print("Pop!_OS uses a symbolic link for the os-release file, this causes issues \
+            and can be fixed by converting to a hard link")
+        print("Attempting to change symlink to hard link for /etc/os-release -> /etc/pop-os/os-release")
```
Comment (bobslept) [inline auto_cpufreq/core.py:90]: I would just remove the else block below

**Comment by tyleraharrison** [comment]
I too have never been able to do code revisions in pull requests, been meaning to learn. Will revise

**Comment by tyleraharrison** [comment]
With this modification users are still able to continue even though the OS name was never read if they press Y. Is that intended behavior?

**Comment by tyleraharrison** [comment]
dist_name will not be initialized and cause issues at `elif dist_name in ["arch", "manjaro", "garuda"]:`

**Comment by bobslept** [comment]
No now your `Continue? [y/N]` is part of the `if os.path.exists("/etc/pop-os/os-release"):` code.
So if the file exists, you will ask them to continue, if that file did not exists, it will go to the `print("Aborting...")`

**Comment by tyleraharrison** [comment]
I was just looking at that, didn't notice it got indented. Making the change now :+1:

**Comment by bobslept** [comment]
Great I think we are there! Now we just wait for @AdnanHodzic to review it one last time and then probably merge it when he is happy! 

I will already say thank you for your contribution to auto-cpufreq!  :+1:

**Comment by tyleraharrison** [comment]
Glad I could help, this tool has served me well on my new Framework laptop so I thought I'd give back.

File: auto_cpufreq/core.py [hunk 4]
```diff
@@ -78,7 +78,26 @@ def get_config(config_file=''):
...
+
+        yN = input("Continue? [y/N] ")
+        if yN.lower() == "y":
+            # Backup /etc/os-release
+            os.system("sudo mv /etc/os-release /etc/os-release-backup")
+            # Create hard link to /etc/os-release
+            os.system("sudo ln /etc/pop-os/os-release /etc/os-release")
```
Comment (bobslept) [inline auto_cpufreq/core.py:97]:
```suggestion
        if yN.lower() == "y":
            # Backup /etc/os-release
            os.system("sudo mv /etc/os-release /etc/os-release-backup")
            # Create hard link to /etc/os-release
            os.system("sudo ln /etc/pop-os/os-release /etc/os-release")
        else:
            # choose N so abort
            print("Operation aborted")
            sys.exit(1)
```

**Review by bobslept — COMMENTED** [review]
Well I've just spotted that we do nothing, when we chose N option. So maybe it's a good idea to actually handle it.

**Comment by tyleraharrison** [comment]
Did not mean to hit resolve, let me open the code real quick

**Comment by bobslept** [comment]
I don't really know if I did that right. But the thing is. If the user choose N. we need also nicely print abort and exit.

**Comment by tyleraharrison** [comment]
Is there a clean way to do that without just adding an identical abort block?

**Comment by tyleraharrison** [comment]
```
    yN = input("Continue? [y/N] ")
        if yN.lower() == "y":
            # Backup /etc/os-release
            os.system("sudo mv /etc/os-release /etc/os-release-backup")
            # Create hard link to /etc/os-release
            os.system("sudo ln /etc/pop-os/os-release /etc/os-release")
        else:
            print("Aborting...")
            sys.exit(1)
    else:
        print("Aborting...")
        sys.exit(1)
```
        
Seems a little dirty but it gets the job done

**Comment by tyleraharrison** [comment]
image

**Comment by bobslept** [comment]
Yeah I see. Not really know a nicer way of doing this at the moment. That's why I added another print message. Something like "Operation aborted by user." Would make more sense on that print maybe.

**Comment by tyleraharrison** [comment]
That should do it

File: auto_cpufreq/core.py [hunk 5]
```diff
@@ -78,7 +78,28 @@ def get_config(config_file=''):
...
+except PermissionError:
+    # Current work-around for distros like Pop!_OS where symlink causes permission issues
+    print("Warning: Cannot get distro name")
+    if os.path.exists("/etc/pop-os/os-release"):
+        print("Pop!_OS detected")
+        print("Pop!_OS uses a symbolic link for the os-release file, this causes issues and can be fixed by converting to a hard link")
+        print("Attempting to change symlink to hard link for /etc/os-release -> /etc/pop-os/os-release")
+
+        yN = input("Continue? [y/N] ")
+        if yN.lower() == "y":
```
Comment (bobslept) [inline auto_cpufreq/core.py:92]:
```suggestion
        if yN.lower() == "y":
            print("Creating hard link for /etc/os-release")
```

**Review by bobslept — COMMENTED** [review]
What do you think as last change. So actually printing it's doing something after choosing "y"

File: auto_cpufreq/core.py (hunk 5 above)
Comment (tyleraharrison) [reply auto_cpufreq/core.py:92]: After pressing "commit changes" it does nothing 🤦‍♂️

**Comment by bobslept** [comment]
Wow that's was fast! Well I think it's great! Thank you sir :+1:

**Comment by tyleraharrison** [comment]
Yeah now I understand how the "commit changes" works, just didn't see the indicator for a bit on my side

**Comment by AdnanHodzic** [comment]
Looks great, thank you for your contribution @tyleraharrison and thank you for your assistance reviewing this PR @bobslept!

**Pull request merged**